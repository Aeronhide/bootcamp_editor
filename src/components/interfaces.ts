export interface ShapeItem {
  id: string
  left: number
  top: number
}

export interface ShapeMap {
  [key: string]: { top: number; left: number; }
}
