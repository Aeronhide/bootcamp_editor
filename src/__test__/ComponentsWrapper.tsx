import { configureStore } from '@reduxjs/toolkit'
import { shapesReducer } from '../reducers/shapes.reducer'
import { Provider } from 'react-redux'
import { DndProvider } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend'

export const AppDndWrapper = (child: JSX.Element) => {
  const store = configureStore({
    reducer: {
      shapesReducer,
    }
  })

  return (
    <Provider store={store}>
      <DndProvider backend={HTML5Backend} options={{ enableMouseEvents: true }}>
        {child}
      </DndProvider>
    </Provider>
  )
}
