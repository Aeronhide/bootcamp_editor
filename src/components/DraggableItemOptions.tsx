import * as React from 'react'
import { useDispatch } from 'react-redux'
import { UpdateShapeAction } from '../actions/shape.actions';

interface DraggableItemOptionsProps {
  id: string,
  scale: number,
  zIndex: number
}

export function DraggableItemOptions(props: DraggableItemOptionsProps): JSX.Element {
  let { id, scale, zIndex } = props;
  const dispatch = useDispatch();

  const expandShape = () => {
    dispatch(UpdateShapeAction({ id, scale: scale += 0.1, zIndex }))
  }

  const shrinkShape = () => {
    dispatch(UpdateShapeAction({ id, scale: scale -= 0.1, zIndex }))
  }

  const bringFrontShape = () => {
    dispatch(UpdateShapeAction({ id, scale, zIndex: zIndex += 1 }))
  }

  const sendBackShape = () => {
    if (zIndex === 0) return dispatch(UpdateShapeAction({ id, scale, zIndex: 0 }))

    dispatch(UpdateShapeAction({ id, scale, zIndex: zIndex -= 1 }))
  }

  return (
    <div className="options">
      <div className="options_single">
        <label>Scale</label>
        <div>
          <button className='btn' onClick={shrinkShape}>-</button>
          <span style={{ padding: '0 10px' }}>{scale.toFixed(1)}</span>
          <button className='btn' onClick={expandShape}>+</button>
        </div>
      </div>
      <div className="options_single">
        <label>Z-index</label>
        <div style={{ display: 'flex' }}>
          <button className='btn text-small' onClick={bringFrontShape}>Bring front</button>
          <p className="options_single_z_index">{zIndex.toFixed(0)}</p>
          <button className='btn text-small' onClick={sendBackShape}>Send back</button>
        </div>
      </div>
    </div>
  );
}


