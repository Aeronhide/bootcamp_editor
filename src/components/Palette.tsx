import * as React from "react";
import { DraggableItem } from "./dndComponents";

export function Palette(): JSX.Element {
  return (
    <div className="palette">
      <DraggableItem id="figure1" left={10} top={10} />
      <DraggableItem id="figure2" left={10} top={100} />
      <DraggableItem id="figure3" left={10} top={190} />
    </div>
  );
}


