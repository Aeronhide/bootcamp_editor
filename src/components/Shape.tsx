import * as React from "react";
import { FC } from 'react'
import squareUrl from "../assets/square.svg"
import circleUrl from "../assets/circle.svg"
import triangleUrl from "../assets/triangle.svg"

interface ShapeProp {
  [key: string]: string
}

interface ShapeProps {
  shapeType: string,
  scale: number,
  editing: boolean,
  setEditing(editing: boolean): void
}

export const Shape: FC<ShapeProps> = (props) => {
  const { shapeType, scale, setEditing, editing } = props

  const shapePath: ShapeProp = {
    "square": squareUrl,
    "circle": circleUrl,
    "triangle": triangleUrl
  }

  const shapesColor: ShapeProp = {
    "square": "#FFFF00",
    "triangle": "#000",
    "circle": "#fff",
  }

  const doubleClick = () => {
    setEditing(!editing)
  }
  return (
    <div className={`shape ${shapeType}`} data-testid="shape" onDoubleClick={doubleClick} style={{ transform: `scale(${scale})`, border: editing ? '1px dashed #0000FF' : 'none' }}>
      <img className="shape_img" src={shapePath[shapeType]} alt="shape" />
    </div >
  )
}
