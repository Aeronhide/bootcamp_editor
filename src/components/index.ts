export { Palette } from "./Palette";
export { Workspace } from "./Workspace";
export { Container } from './Container'
export { DraggableItem } from './dndComponents'
export { Shape } from './Shape'
