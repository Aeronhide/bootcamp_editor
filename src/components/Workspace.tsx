import * as React from 'react'
import { useCallback } from 'react'
import type { XYCoord } from 'react-dnd'
import { useDrop } from 'react-dnd'
import { useSelector } from 'react-redux'
import { useDispatch } from 'react-redux'
import { AddToWorkSpaceAction, ResetEditShapesAction } from '../actions/shape.actions'
import { ShapesTypes as ST } from '../constants/shapesTypes'
import { DraggableItem } from './dndComponents/DraggableItem'
import { DraggableItemOptions } from './DraggableItemOptions'
import { ShapeItemRequired } from './interfaces'
import { RootState } from '../store';

export function Workspace(): JSX.Element {
  const dispatch = useDispatch();
  const reduxState = useSelector((state: RootState) => state.shapesReducer.shapes);

  const moveShape = useCallback(
    (item: ShapeItemRequired) => {
      const { id, left, top, shape, width, height, scale, zIndex, editing } = item;
      dispatch(AddToWorkSpaceAction({ [id]: { left, top, shape, width, height, scale, zIndex, editing } }))
    },
    [reduxState, AddToWorkSpaceAction],
  )

  const [, drop] = useDrop(
    () => ({
      accept: ST.COMMON,
      drop(item: ShapeItemRequired, monitor) {
        const delta = monitor.getDifferenceFromInitialOffset() as XYCoord
        const left = Math.round(item.left + delta.x)
        const top = Math.round(item.top + delta.y)
        moveShape({ ...item, left, top })
      },
      collect: (monitor) => ({
        isOver: !!monitor.isOver()
      })
    }),
  )

  const resetEditShapes = (e: React.MouseEvent<HTMLElement>) => {
    if ((e.target as Element).className === "canvas") return dispatch(ResetEditShapesAction())
  }

  return (
    <div className="workspace">
      <div ref={drop} className="canvas" onClick={resetEditShapes} title="dropSource">
        {Object.keys(reduxState).map((shapeKey) => {
          return (
            <DraggableItem
              key={shapeKey}
              id={shapeKey}
              {...reduxState[shapeKey]}
            />
          )
        })
        }
      </div>
      {Object.keys(reduxState).map((shapeKey) => {
        if (reduxState[shapeKey].editing) {
          return (
            <DraggableItemOptions
              key={"options" + shapeKey}
              id={shapeKey}
              scale={reduxState[shapeKey].scale}
              zIndex={reduxState[shapeKey].zIndex}
            />
          )
        }
      })}
    </div>
  );
}


