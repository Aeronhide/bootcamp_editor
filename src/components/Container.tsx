import * as React from 'react'
import { Palette } from './Palette'
import { Workspace } from './Workspace'

export const Container = (): JSX.Element => {
  return (
    <div data-testid="container">
      <Palette />
      <Workspace />
    </div>
  )
}
