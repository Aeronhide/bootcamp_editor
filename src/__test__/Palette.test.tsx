import { render, screen, fireEvent } from '@testing-library/react';
import { Palette } from '../components/Palette';
import { AppDndWrapper } from './ComponentsWrapper';

describe('Palette', () => {
  it('must contain given count of figures', () => {
    const { container } = render(
      AppDndWrapper(<Palette />)
    );
    const shapes = container.getElementsByClassName('shape');
    expect(shapes.length).toBe(3)
  })
})
