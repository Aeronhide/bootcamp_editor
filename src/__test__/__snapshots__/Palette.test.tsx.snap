// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Palette must be equal to snapshot 1`] = `
<div>
  <div
    class="palette"
    title="palette"
  >
    <div
      class="palette_shape_wrapper"
    >
      <div
        class="box"
        draggable="true"
        style="position: absolute; background: transparent; cursor: move; left: 10px; top: 10px; width: 50px; height: 50px; z-index: 10;"
        title="dragSource"
      >
        <div
          class="shape square"
          data-testid="shape"
          style="transform: scale(1);"
        >
          <img
            alt="shape"
            class="shape_img"
            src="square.svg"
          />
        </div>
      </div>
    </div>
    <div
      class="palette_shape_wrapper"
    >
      <div
        class="box"
        draggable="true"
        style="position: absolute; background: transparent; cursor: move; left: 10px; top: 10px; width: 50px; height: 50px; z-index: 10;"
        title="dragSource"
      >
        <div
          class="shape triangle"
          data-testid="shape"
          style="transform: scale(1);"
        >
          <img
            alt="shape"
            class="shape_img"
            src="triangle.svg"
          />
        </div>
      </div>
    </div>
    <div
      class="palette_shape_wrapper"
    >
      <div
        class="box"
        draggable="true"
        style="position: absolute; background: transparent; cursor: move; left: 10px; top: 10px; width: 50px; height: 50px; z-index: 10;"
        title="dragSource"
      >
        <div
          class="shape circle"
          data-testid="shape"
          style="transform: scale(1);"
        >
          <img
            alt="shape"
            class="shape_img"
            src="circle.svg"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
